draft-yasskin-http-origin-signed-responses.md

```diff
@@ -0,0 +1,316 @@
+
+stand_alone: yes
+pi: [comments, sortrefs, strict, symrefs, toc]
+
+author:
+ -
+    name: Jeffrey Yasskin
+    organization: Google
+    email: jyasskin@chromium.org
+
+informative:
+  SRI: W3C.REC-SRI-20160623
+
+--- abstract
+
+This document explores how a server can send particular responses that are
+authoritative for an origin, when the server itself is not authoritative for
+that origin. For now, it focuses on the constraints covering any such mechanism.
+
+--- note_Note_to_Readers
+
+Discussion of this draft takes place on the HTTP working group mailing list
+(ietf-http-wg@w3.org), which is archived
+at <https://lists.w3.org/Archives/Public/ietf-http-wg/>.
+
+The source code and issues list for this draft can be found
+in <https://github.com/WICG/webpackage>.
+
+--- middle
+
+# Introduction
+
+When
+I
+[presented Web Packaging to DISPATCH](https://datatracker.ietf.org/doc/minutes-99-dispatch/),
+folks thought it would make sense to split it into a way to sign individual HTTP
+responses as coming from a particular origin, and separately a way to bundle a
+collection of HTTP responses. This document explores the constraints on any
+method of signing HTTP responses and briefly sketches a possible solution to the
+constraints.
+
+# Terminology
+
+Author
+: The entity that controls the server for a particular origin {{?RFC6454}}. The
+  author can get a CA to issue certificates for their private keys and can run a
+  TLS server for their origin.
+
+Intermediate
+: An entity that fetches signed HTTP responses from an author or another
+  intermediate and forwards them to another intermediate or a client.
+
+Client
+: An entity that uses a signed HTTP response and needs to be able to prove that
+  the author vouched for it as coming from its claimed origin.
+
+# Use cases
+
+## PUSHed subresources {#uc-pushed-subresources}
+
+To reduce round trips, a server might use HTTP/2 PUSH to inject a subresource
+from another server into the client's cache. If anything about the subresource
+is expired or can't be verified, the client would fetch it from the original
+server.
+
+For example, if `https://example.com/index.html` includes
+
+~~~html
+<script src="https://jquery.com/jquery-1.2.3.min.js">
+~~~
+
+Then to avoid the need to look up and connect to `jquery.com` in the critical
+path, `example.com` might PUSH that resource ({{?RFC7540}}, section 8.2), signed
+by `jquery.com`.
+
+## Explicit use of a CDN for subresources {#uc-explicit-cdn}
+
+In order to speed up loading but still maintain control over its content, an
+HTML page in a particular origin `O.com` could tell clients to load its
+subresources from an intermediate CDN that's not authoritative, but require that
+those resources be signed by `O.com` so that the CDN couldn't modify the
+resources.
+
+~~~html
+<img logicalsrc="https://O.com/img.png"
+     physicalsrc="https://cdn.com/O.com/img.png">
+~~~
+
+This could be used for some of the same purposes as SRI ({{uc-sri}}).
```

Comment:
The overall use case makes sense, but enumerating CDNs in markup is painful. Could this be elevated to a document-level opt-in (e.g. via response header or meta directive) that notifies UA that content on this origin can also be fetched from cdn.com? Specifically, say I detect that user is coming from Geo-region X, or mobile network Y, and I have edge caches optimized for each of those... I'd like to emit a header that notifies UA that it can fetch from X or Y; rewriting HTML is super painful.

Reply:
FWIW, I'm skeptical that's a scalable way to do this, or even required one for this use case.. What I'm describing is another iteration on base tag.